roms_tools/setup/cdr_forcing.py

```diff
@@ -21,171 +18,210 @@
-        Tracer concentrations are extended accordingly by duplicating endpoint values.
+class CDRForcingDatasetBuilder:
+    """Constructs the xarray `Dataset` to be saved as NetCDF."""
 
-        If `times` is empty, [start_time, end_time] is used with zero fluxes and duplicated tracer values.
-        """
-        if not self.times:
-            self.times = [self.start_time, self.end_time]
-            self.volume_fluxes = [self.volume_fluxes, self.volume_fluxes]
-            self.tracer_concentrations = {
-                tracer: [vals, vals]
-                if not isinstance(vals, list)
-                else [vals[0], vals[0]]
-                for tracer, vals in self.tracer_concentrations.items()
+    def __init__(self, releases, model_reference_date, release_type):
+        self.releases = releases
+        self.model_reference_date = model_reference_date
+        self.release_type = release_type
+
+    def build(self) -> xr.Dataset:
+        all_times = [np.array(r.times, dtype="datetime64[ns]") for r in self.releases]
+        union_times = np.unique(np.concatenate(all_times))
+        union_rel_times = [
+            convert_to_relative_days(t, self.model_reference_date) for t in union_times
+        ]
+
+        ds = xr.Dataset()
+        ds["time"] = ("time", union_times)
+        ds["cdr_time"] = ("time", union_rel_times)
+        ds["cdr_lon"] = ("ncdr", [r.lon for r in self.releases])
+        ds["cdr_lat"] = ("ncdr", [r.lat for r in self.releases])
+        ds["cdr_dep"] = ("ncdr", [r.depth for r in self.releases])
+        ds["cdr_hsc"] = ("ncdr", [r.hsc for r in self.releases])
+        ds["cdr_vsc"] = ("ncdr", [r.vsc for r in self.releases])
+        ds = ds.assign_coords(
+            {"release_name": (["ncdr"], [r.name for r in self.releases])}
+        )
+
+        # Assign attributes
+        attr_map = {
+            "time": {"long_name": "absolute time"},
+            "cdr_time": {
+                "long_name": f"relative time: days since {self.model_reference_date}",
+                "units": "days",
+            },
+            "release_name": {"long_name": "Name of release"},
+            "cdr_lon": {
+                "long_name": "Longitude of CDR release",
+                "units": "degrees east",
+            },
+            "cdr_lat": {
+                "long_name": "Latitude of CDR release",
+                "units": "degrees north",
+            },
+            "cdr_dep": {"long_name": "Depth of CDR release", "units": "meters"},
+            "cdr_hsc": {
+                "long_name": "Horizontal scale of CDR release",
+                "units": "meters",
+            },
+            "cdr_vsc": {
+                "long_name": "Vertical scale of CDR release",
+                "units": "meters",
+            },
+        }
+        for var, attrs in attr_map.items():
+            ds[var].attrs.update(attrs)
+
+        if self.release_type == VolumeRelease:
+            ds = add_tracer_metadata_to_ds(
```

Comment:
Consider pulling each of your big cases (`if self.release_type == xxx`) into it's own method to simplify testing of each subsection and shorten this monster!